### PR TITLE
Ticket2764 create ioc generator

### PR DESCRIPTION
**Pull request by AdrianPotter**
Create a device from scratch. Includes:

- OPI
- Support submodule
- IOC
- Emulator
- IOC tests

run as `python IBEX_device_generator.py --name=[ioc_name] --ticket=[ticket_number]`

https://github.com/ISISComputingGroup/IBEX/issues/2764

### CODE REVIEW

File: utils/git_utils.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        logging.info("Preparing new branch, {}, for repo {}".format(branch, self._repo.working_tree_dir))
+        if self._repo.is_dirty() and ask_do_step(
+                "Repository {} is dirty, clean it? (Prompts will be given for subsequent steps)"
+                "".format(self._repo.working_tree_dir)):
+            try:
+                if ask_do_step("Perform a hard reset on the repository"):
+                    self._repo.git.reset("HEAD", hard=True)
+                if ask_do_step("Perform git clean -fd on the repository"):
+                    self._repo.git.clean(f=True, d=True)
```
Comment (Tom-Willemsen) [inline utils/git_utils.py]: I'm still hesitant to let the user do a `clean` and `reset`... I wonder if `git stash --include-untracked` would be sufficient here? The advantages that I can see are (1) it doesn't touch ignored (i.e. built) files and (2) it is reversible if the user makes a mistake.

**Review by Tom-Willemsen — CHANGES_REQUESTED** [review]
Looks like it's mostly there.

Launching the IOC I get a couple of weird errors:

```
Warning: IOC is booting with TOP = "C:\Instrument\Apps\EPICS\support\EPICS_V4\master\"
          but was built with TOP = "C:/Instrument/Apps/EPICS/ioc/master/ADRIAN"
```

and

```
No such file or directory dbRead opening file ../../db/ADRIAN-IOC-01.db
```

(it's substituted in the wrong name for the DB file)

I also noticed a couple of things which would be nice to do, but don't *have* to be done if they're hard:

- The st.cmd is missing a `epicsEnvSet "STREAM_PROTOCOL_PATH" "$(AMINT2L)/data"` line
- The support module is not built
- Makefile entry not added in ioc directory (to long list of DIRS_TO_BUILD). It looks like there is logic for this?
- Makefile entry not added to support directory, it also looks like there was logic for this?

Finally,
- Could you add a wiki page about it describing how to use it, what it does for you, and what it doesn't do for you
- Could you add release notes

**Comment by AdrianPotter** [comment]
- I've added a more interactive selection of the branch cleaning option. The choice you suggested is now available. The existing choices remain but have explicit implications attached to them and you have to do an extra level of confirmation after selecting them.
- I intentionally didn't include the ticket on the release notes as it's not a change that's included in the release or that has any impact on user functionality.
- The `st.cmd` should include all our usual stuff about Freeport and `TESTDEVSIM`. If it looks too plain (e.g. no stream protocol path) then it hasn't picked up the correct template. You need to pick up the attached PR from EPICS-base and have built the submodule with the new template
- Add make command to support submodule
- Fix a bug in adding the new dirs to their respective makefiles (+tests)
- I cannot reproduce the issues with `TOP` I've made the IOC from scratch and started the IOC from the console and it does not display the error. I've linked the branches created from the script for confirmation, along with the IOC logs for comparison
- I've updated the README.MD on the device generator repo for information about how to run

**Comment by AdrianPotter** [comment]
https://github.com/ISISComputingGroup/EPICS-ioc/tree/Ticket9005_Add_IOC_ADRIAN
https://github.com/ISISComputingGroup/EPICS/tree/Ticket9005_Add_IOC_ADRIAN
https://github.com/ISISComputingGroup/EPICS-Adrian/tree/Ticket9005_Add_IOC_ADRIAN
https://github.com/ISISComputingGroup/ibex_gui/tree/Ticket9005_Add_IOC_ADRIAN
https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/tree/Ticket9005_Add_IOC_ADRIAN
https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/tree/Ticket9005_Add_IOC_ADRIAN

**Comment by AdrianPotter** [comment]
[ADRIAN_01-20171127.log](https://github.com/ISISComputingGroup/IBEX_device_generator/files/1506967/ADRIAN_01-20171127.log)

**Review by Tom-Willemsen — APPROVED** [review]

**Pull request merged**